drivers/adc/adc_gd32.c

```diff
@@ -0,0 +1,396 @@
+/*
+ * Copyright (c) 2022 BrainCo Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_adc
+
+#include <errno.h>
+#include <drivers/pinctrl.h>
+#include <drivers/adc.h>
+#include <soc.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_gd32, CONFIG_ADC_LOG_LEVEL);
+
+#if defined(CONFIG_SOC_SERIES_GD32F3X0)
+/* Cancle original registers definitions */
+#undef ADC_STAT
+#undef ADC_CTL0
+#undef ADC_CTL1
+#undef ADC_SAMPT0
+#undef ADC_SAMPT1
+#undef ADC_RSQ2
+#undef ADC_RDATA
+/* Registers definitions with dummy parameter */
+#define ADC_STAT(_)	REG32(ADC + 0x00000000U)
+#define ADC_CTL0(_)	REG32(ADC + 0x00000004U)
+#define ADC_CTL1(_)	REG32(ADC + 0x00000008U)
+#define ADC_SAMPT0(_)	REG32(ADC + 0x0000000CU)
+#define ADC_SAMPT1(_)	REG32(ADC + 0x00000010U)
+#define ADC_RSQ2(_)	REG32(ADC + 0x00000034U)
+#define ADC_RDATA(_)	REG32(ADC + 0x0000004CU)
+#endif
+
+#define SPT_WIDTH	3U
+#define SAMPT1_SIZE	10U
+
+#if defined(CONFIG_SOC_SERIES_GD32F4XX)
+#define SMP_TIME(x)	ADC_SAMPLETIME_##x
+
+static const uint16_t acq_time_tbl[8] = {3, 15, 28, 56, 84, 112, 144, 480};
+static const uint32_t table_samp_time[] = {
+	SMP_TIME(3),
+	SMP_TIME(15),
+	SMP_TIME(28),
+	SMP_TIME(56),
+	SMP_TIME(84),
+	SMP_TIME(112),
+	SMP_TIME(144),
+	SMP_TIME(480)
+};
+#else
+#define SMP_TIME(x)	ADC_SAMPLETIME_##x##POINT5
+
+static const uint16_t acq_time_tbl[8] = {2, 8, 14, 29, 42, 56, 72, 240};
+static const uint32_t table_samp_time[] = {
+	SMP_TIME(1),
+	SMP_TIME(7),
+	SMP_TIME(13),
+	SMP_TIME(28),
+	SMP_TIME(41),
+	SMP_TIME(55),
+	SMP_TIME(71),
+	SMP_TIME(239)
+};
+#endif
+
+struct adc_gd32_config {
+	uint32_t reg;
+	uint32_t rcu_periph_clock;
+	uint8_t channels;
+	const struct pinctrl_dev_config *pcfg;
+	void (*irq_config_func)(void);
+};
+
+struct adc_gd32_data {
+	struct adc_context ctx;
+	const struct device *dev;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+};
+
+/* Use a flag to protect the shared irq handler initialization. */
+static bool init_irq = true;
+
+static void adc_gd32_isr(const struct device *dev)
+{
+	struct adc_gd32_data *data = dev->data;
+	const struct adc_gd32_config *cfg = dev->config;
+
+	if (ADC_STAT(cfg->reg) & ADC_STAT_EOC) {
+		*data->buffer++ = ADC_RDATA(cfg->reg);
+
+		/* Disable EOC interrupt. */
+		ADC_CTL0(cfg->reg) &= ~ADC_CTL0_EOCIE;
+		/* Clear EOC bit. */
+		ADC_STAT(cfg->reg) &= ~ADC_STAT_EOC;
+
+		adc_context_on_sampling_done(&data->ctx, dev);
+	}
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_gd32_data *data =
+		CONTAINER_OF(ctx, struct adc_gd32_data, ctx);
```

Comment:
The column width has been increased to 100, I think you should be able to fit this into a single line